les/clientpool.go

```diff
@@ -0,0 +1,591 @@
+	if newCapacity > f.capacityLimit || newCount > f.countLimit {
+		var (
+			kickList     []*clientInfo
+			kickPriority int64
+		)
+		f.connectedQueue.MultiPop(func(data interface{}, priority int64) bool {
+			c := data.(*clientInfo)
+			kickList = append(kickList, c)
+			kickPriority = priority
+			newCapacity -= c.capacity
+			newCount--
+			return newCapacity > f.capacityLimit || newCount > f.countLimit
+		})
+		if newCapacity > f.capacityLimit || newCount > f.countLimit || (e.balanceTracker.estimatedPriority(now+mclock.AbsTime(connectedBias), false)-kickPriority) > 0 {
+			// reject client
+			for _, c := range kickList {
+				f.connectedQueue.Push(c)
+			}
+			log.Debug("Client rejected", "address", address, "id", idStr)
+			return nil
+		}
+		// accept new client, drop old ones
+		for _, c := range kickList {
+			f.dropClient(c, now, true)
+		}
+	}
+	// client accepted, finish setting it up
+	if nb != nil {
+		delete(f.negBalanceMap, address)
+		f.negBalanceQueue.Remove(nb.queueIndex)
+	}
+	if e.priority {
+		e.balanceTracker.addCallback(balanceCallbackZero, 0, func() { f.balanceExhausted(id) })
+	}
+	f.connectedMap[id] = e
+	f.connectedQueue.Push(e)
+	f.connectedCapacity += e.capacity
+	if e.capacity != f.freeClientCap {
+		e.peer.updateCapacity(e.capacity)
+	}
+	log.Debug("Client accepted", "address", address)
+	return e
+}
+
+// unregisterPeer implements peerSetNotify
+func (f *clientPool) unregisterPeer(p *peer) {
+	f.disconnect(p)
+}
+
+// disconnect should be called when a connection is terminated. If the disconnection
+// was initiated by the pool itself using disconnectFn then calling disconnect is
+// not necessary but permitted.
+func (f *clientPool) disconnect(p clientPeer) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	if f.closed {
+		return
+	}
+	address := p.freeClientId()
+	id := p.ID()
+	// Short circuit if the peer hasn't been registered.
+	e := f.connectedMap[id]
+	if e == nil {
+		log.Debug("Client not connected", "address", address, "id", peerIdToString(id))
+		return
+	}
+	f.dropClient(e, f.clock.Now(), false)
+}
+
+// dropClient removes a client from the connected queue and finalizes its balance.
+// If kick is true then it also initiates the disconnection.
+func (f *clientPool) dropClient(e *clientInfo, now mclock.AbsTime, kick bool) {
+	if _, ok := f.connectedMap[e.id]; !ok {
+		return
+	}
+	f.finalizeBalance(e, now)
+	f.connectedQueue.Remove(e.queueIndex)
+	delete(f.connectedMap, e.id)
+	f.connectedCapacity -= e.capacity
+	if kick {
+		log.Debug("Client kicked out", "address", e.address)
+		f.removePeer(e.id)
+	} else {
+		log.Debug("Client disconnected", "address", e.address)
+	}
+}
+
+// finalizeBalance stops the balance tracker, retrieves the final balances and
+// stores them in posBalanceQueue and negBalanceQueue
+func (f *clientPool) finalizeBalance(c *clientInfo, now mclock.AbsTime) {
+	c.balanceTracker.stop(now)
+	pos, neg := c.balanceTracker.getBalance(now)
+	pb := f.getPosBalance(c.id)
```

Comment:
Can we also store the `posBalance` into database if a client is dropped. Currently we will maintain the positive balance in Queue(which is cache) and flush items if the size of queue exceeds the limit(of course we will flush all positive balances when clientpool is closed). So if the Geth node is crashed, we can lose a lot of positive balances

Reply:
We save them immediately when the balance is increased but I think we can also save at disconnection. Will add it soon.